File: mini_ep2/crivo_turbo.c
// miniEP1
// Giovana Vieira de Morais, 11921908
// gcc crivo.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void main(int argc, char *argv[]) {
	int N, *crivo;
	char *ptr;
	unsigned long int primos = 0, primos_especiais = 0;
	
	// criar um vetor com metade do tamanho pra economizar memória
	N = 1 << strtol(argv[1], &ptr, 10);
	crivo = (int*) malloc(sizeof(int) * (N+1));

	for (unsigned long int i = 2; i <= N; i++) 
		crivo[i] = 1;

	for (unsigned long int i = 2; i <= N; i++) { 
		if (crivo[i]) {
			primos += 1;

			if (i % 4 != 3) 
				primos_especiais += 1;

			for (unsigned long int j = i*2; j <= N; j += i) 
				crivo[j] = 0;
		}
	}
	printf("%lu %lu\n", primos, primos_especiais);
	free(crivo);
}

File: mini_ep2/crivo_turbo.py
# miniEP1
# Giovana Vieira de Morais, 11921908
# python3 crivo.py
import sys

N = 1 << int(sys.argv[1])
primos = 0
primos_especiais = 0
crivo = [True]*(N+1)

for i in range(2, N+1):
    if crivo[i]:
        primos = primos + 1

        if i % 4 != 3:
            primos_especiais = primos_especiais + 1
        for j in range(i*2, N+1, i):
            crivo[j] = False

print(primos, primos_especiais)
